prometheus/graphite/bridge.go

```diff
@@ -0,0 +1,244 @@
+type Logger interface {
+	Printf(v ...interface{})
+}
+
+// NewBridge returns a pointer to a new Bridge struct.
+func NewBridge(c *Config) (*Bridge, error) {
+	b := &Bridge{}
+
+	if c.URL == "" {
+		return nil, errors.New("missing URL")
+	}
+	b.url = c.URL
+
+	if c.Gatherer == nil {
+		b.g = prometheus.DefaultGatherer
+	} else {
+		b.g = c.Gatherer
+	}
+
+	if c.Logger != nil {
+		b.logger = c.Logger
+	}
+
+	if c.Prefix != "" {
+		b.prefix = c.Prefix
+	}
+
+	var z time.Duration
+	if c.Interval == z {
+		b.interval = defaultInterval
+	} else {
+		b.interval = c.Interval
+	}
+
+	if c.Timeout == z {
+		b.timeout = defaultInterval
+	} else {
+		b.timeout = c.Timeout
+	}
+
+	b.errorHandling = c.ErrorHandling
+
+	return b, nil
+}
+
+// Run starts the event loop that pushes Prometheus metrics to Graphite at the
+// configured interval.
+func (b *Bridge) Run(ctx context.Context) {
+	ticker := time.NewTicker(b.interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if err := b.Push(); err != nil {
+				if b.logger != nil {
+					b.logger.Printf("%v", err)
+				}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// Push pushes Prometheus metrics to the configured Graphite server.
+func (b *Bridge) Push() error {
+	mfs, err := b.g.Gather()
+	if err != nil {
+		switch b.errorHandling {
+		case AbortOnError:
+			return err
+		case ContinueOnError:
+			if b.logger != nil {
+				b.logger.Printf("continue on error: %v", err)
+			}
+		}
+	}
+
+	conn, err := net.DialTimeout("tcp", b.url, b.timeout)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	_, err = writeMetrics(conn, mfs, b.prefix, model.Now())
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+func writeMetrics(w io.Writer, mfs []*dto.MetricFamily, prefix string, now model.Time) (int, error) {
+	vec := expfmt.ExtractSamples(&expfmt.DecodeOptions{
+		Timestamp: now,
+	}, mfs...)
+
+	var total int
+	for _, s := range vec {
+		n, err := fmt.Fprintf(w, "%s %g %d\n", strings.Join([]string{sanitize(prefix), toMetric(s.Metric)}, "."), s.Value, int64(s.Timestamp)/millisecondsPerSecond)
+		if err != nil {
+			return 0, err
+		}
+		total += n
+	}
+
+	return total, nil
+}
+
+func toMetric(m model.Metric) string {
+	metricName, hasName := m[model.MetricNameLabel]
+	numLabels := len(m) - 1
+	if !hasName {
+		numLabels = len(m)
+	}
+
+	labelStrings := make([]string, 0, numLabels)
+	for label, value := range m {
+		if label != model.MetricNameLabel {
+			labelStrings = append(labelStrings, fmt.Sprintf("%s.%s", sanitize(string(label)), sanitize(string(value))))
+		}
+	}
+
+	switch numLabels {
+	case 0:
+		if hasName {
+			return sanitize(string(metricName))
+		}
+		return ""
+	default:
+		sort.Strings(labelStrings)
+		return fmt.Sprintf("%s.%s", sanitize(string(metricName)), strings.Join(labelStrings, "."))
+	}
+}
+
+var (
+	reInvalidChars       = regexp.MustCompile("[^a-zA-Z0-9_-]")
+	reRepeatedUnderscore = regexp.MustCompile("_{2,}")
+)
+
+func sanitize(s string) string {
```

Comment:
As per the suggestion above, this should take a bufio.Writer and stream directly into it.

Code would be something like:

```Go
prevUnderscore := false

for _, c := range s {
    c = replaceInvalidRune(c)
    if c == '_' {
        if prevUnderscore {
            continue
        }
        prevUnderscore = true
    } else {
        prevUnderscore = false
    }
    if _, err := buf.WriteRune(c); err != nil {
        return err
    }
}

func replaceInvalidRune(rune c) rune {
    // do numerical comparisons to identify invalid chars.
   // See https://github.com/prometheus/common/blob/master/model/metric.go#L92-L95
}
```